[release] merging release-2023-02-09 into master (#291)
gtk-rs/gtk-rs.github.io: _posts/2023-02-09-new-release.md

 * [gio: Add a GioInfaliableFuture](https://github.com/gtk-rs/gtk-rs-core/pull/849)
 * [Add a changelog file](https://github.com/gtk-rs/gtk-rs-core/pull/850)
 * [glib: Implement `IntoGlibPtr` for `Option<T>`](https://github.com/gtk-rs/gtk-rs-core/pull/853)
 * [glib: Use actual function name in `glib::BoolError` and include function name/source file/line number in structured logs](https://github.com/gtk-rs/gtk-rs-core/pull/852)
 * [glib: Add `GStr::from_ptr_lossy()` and `GString::from_ptr_lossy()` and implement `From<GStr>` and `From<GString>` for `Cow<GStr>`](https://github.com/gtk-rs/gtk-rs-core/pull/855)
 * [glib: Implement `GStringBuilder` as `BoxedInline` to avoid a useless additional heap allocation](https://github.com/gtk-rs/gtk-rs-core/pull/856)
 * [glib: Minor `GStringBuilder` improvements](https://github.com/gtk-rs/gtk-rs-core/pull/857)
 * [Group imports and use prelude](https://github.com/gtk-rs/gtk-rs-core/pull/854)
 * [glib: Bind more `g_utf8` APIs](https://github.com/gtk-rs/gtk-rs-core/pull/813)
 * [gio: fix clippy lints for rust 1.64](https://github.com/gtk-rs/gtk-rs-core/pull/861)

>  * [gio: add `spawn_blocking()`](https://github.com/gtk-rs/gtk-rs-core/pull/818)

Reviewer: # Spawning of blocking functions on the `glib::MainContext`

`gio::spawn_blocking()` was added, which allows to spawn blocking functions on an internal threadpool and retrieve their results as a `Future` from the `glib::MainContext`. This is useful for doing blocking / CPU intense work in the background and when the results are available to handle them on the application's main thread.

As part of this work, `JoinHandle`-style types were also added to `glib::MainContext::spawn()` and related functions for spawning `Future`s on the main context, and to `glib::ThreadPool` for spawning functions on a custom thread pool and collecting their results.

 * [Suggest kebab-case for the error domain](https://github.com/gtk-rs/gtk-rs-core/pull/874)
 * [gio: simplify async initable](https://github.com/gtk-rs/gtk-rs-core/pull/876)
 * [Improve `glib::collections` API and make it more versatile](https://github.com/gtk-rs/gtk-rs-core/pull/875)
 * [glib: Minor optimization for `List`/`SList` with `Copy` types](https://github.com/gtk-rs/gtk-rs-core/pull/877)
 * [glib: Fix `glib::wrapper!` for `BoxedInline` with generic parameters](https://github.com/gtk-rs/gtk-rs-core/pull/879)
 * [glib: Optimize various from/to `Vec` FFI translation functions](https://github.com/gtk-rs/gtk-rs-core/pull/878)
 * [Use `IntoGlibPtr` instead of `to_glib_full()` in more places](https://github.com/gtk-rs/gtk-rs-core/pull/881)
 * [glib: Add `Value::from_type_unchecked()`](https://github.com/gtk-rs/gtk-rs-core/pull/883)
 * [glib: Use `IntoGStr` trait in a couple of places](https://github.com/gtk-rs/gtk-rs-core/pull/882)
 * [Generate string constants as `&'static GStr` / `[u8]`](https://github.com/gtk-rs/gtk-rs-core/pull/885)
 * [glib: Fix usage of `gformat!` macro if `GString` is not in scope](https://github.com/gtk-rs/gtk-rs-core/pull/886)

>  * [Bump MSRV to 1.64 everywhere](https://github.com/gtk-rs/gtk-rs-core/pull/893)

Reviewer: # Update of minimum supported Rust version (MSRV)

With this release, Rust 1.64 is the minimum version required to compile and use the bindings.

 * [glib: implement `ToGlibPtr<*mut _>` for boxed types](https://github.com/gtk-rs/gtk-rs-core/pull/895)
 * [settings: implement strv setter and getter manually](https://github.com/gtk-rs/gtk-rs-core/pull/899)
 * [gio: use StrV for the simple proxy resolver API](https://github.com/gtk-rs/gtk-rs-core/pull/898)
 * [glib: Add `ParamSpec::is()` helper function](https://github.com/gtk-rs/gtk-rs-core/pull/907)
 * [gdk-pixbuf: Trust return value nullability](https://github.com/gtk-rs/gtk-rs-core/pull/905)
 * [glib: Deprecate `ObjectSubclass::instance()` / `from_instance()` in favour of the shorter versions](https://github.com/gtk-rs/gtk-rs-core/pull/906)
 * [build-tools: Fix reporting of errors](https://github.com/gtk-rs/gtk-rs-core/pull/904)
 * [Do not hard-code path separator](https://github.com/gtk-rs/gtk-rs-core/pull/911)
 * [glib: Implement enum paramspec builder variant that builds the default value automatically](https://github.com/gtk-rs/gtk-rs-core/pull/909)
 * [glib: Deprecate paramspec `new()` functions in favour of the builder](https://github.com/gtk-rs/gtk-rs-core/pull/908)

>  * [glib: Add new object constructor for constructing an object with default property values](https://github.com/gtk-rs/gtk-rs-core/pull/910)

Reviewer: # New `glib::Object` constructors

The old `Object::new()` constructor that took a slice of property name / value pairs does not exist anymore. Instead `Object::new()` takes no parameters anymore and creates a new object with the default property values.

For creating a new object with non-default property values, `Object::builder()` can be used which provides a much nicer API and also allows for better code generation and fewer value copies.

```rust
// before
let obj = glib::Object::new::<MyObject>(&[("a", &123i32), ("b", &"some string")]);
// after
let obj = glib::Object::builder::<MyObject>()
    .property("a", 123i32)
    .property("b", "some string")
    .build();
```

 * [glib: Add `MainContext::spawn_from_within()` for spawning non-`Send` futures from another thread](https://github.com/gtk-rs/gtk-rs-core/pull/901)
 * [mark suboptimal `Object` constructors as deprecated](https://github.com/gtk-rs/gtk-rs-core/pull/914)
 * [glib: Implement more From traits for StrV](https://github.com/gtk-rs/gtk-rs-core/pull/919)
 * [build-tools: Allow passing multiple source dirs to `compile_resources`](https://github.com/gtk-rs/gtk-rs-core/pull/917)
 * [gio: use GStr for the manual extension point implenentation](https://github.com/gtk-rs/gtk-rs-core/pull/920)
 * [glib: Implement various traits on `GStr` manually](https://github.com/gtk-rs/gtk-rs-core/pull/921)
 * [gio: bind GFileDescriptorBased](https://github.com/gtk-rs/gtk-rs-core/pull/915)
 * [glib: Manually implement `TimeZone::adjust_time()` instead of ignoring it](https://github.com/gtk-rs/gtk-rs-core/pull/924)
 * [ci: Only deny clippy warnings with stable clippy](https://github.com/gtk-rs/gtk-rs-core/pull/925)

>  * [application: Return ExitCode](https://github.com/gtk-rs/gtk-rs-core/pull/880)

Reviewer: # `gio::Application::run()` returns an `ExitCode`

Instead of returning a plain `i32`, `gio::Application::run()` now returns an `ExitCode` type. This can be directly returned from the `main()` function as it implements the `Termination` trait and by that makes it much easier to correctly propagate application exit codes.

>  * [Properties macro](https://github.com/gtk-rs/gtk-rs-core/pull/494)

Reviewer: @GuillaumeGomez This one should probably be at the top

# New `#[properties]` macro for making usage of properties in `glib::Object` subclasses easier

The `glib::Properties` macro is now available! It can be used to define object properties from a struct definition.
The macro will:
  - define all the necessary ParamSpecs, with the required flags
  - generate, for each property, a default getter/setter on the wrapper type
  - generate, for each property, a `notify_$property` method on the wrapper type
  - generate the methods `derived_properties` `derived_property` and `derived_set_property`, which can be directly called inside the respective `ObjectImpl` methods.
​
## Quick example
```
#[derive(Properties)]
#[properties(wrapper_type = super::Author)]
pub struct Author {
	#[property(get, set, explicit_notify)]
	name: RefCell<String>,
}
```
​
This is the result of months of work, and will greatly improve the gtk-rs ecosystem, by reducing the boilerplate and making the platform more welcoming. 
​
Be sure to checkout the [documentation](https://gtk-rs.org/gtk-rs-core/stable/latest/docs/glib/derive.Properties.html) for more info.

Reviewer: Also @ranfdev wrote this

 * [widget: support `window_state_event` when subclassing](https://github.com/gtk-rs/gtk3-rs/pull/786)
 * [Skip init assertion for `gdk::set_allowed_backends`](https://github.com/gtk-rs/gtk3-rs/pull/791)
 * [fix windows CI](https://github.com/gtk-rs/gtk3-rs/pull/793)
 * [gtk: implement `From<ResponseType>` for Value](https://github.com/gtk-rs/gtk3-rs/pull/794)
 * [examples: Fix compilation after `gio::SimpleAction` constructor took the state by value](https://github.com/gtk-rs/gtk3-rs/pull/797)
 * [examples: Adapt to glib-build-tools breaking change](https://github.com/gtk-rs/gtk3-rs/pull/798)
 * [Allow subclassing ToggleButton and MenuButton](https://github.com/gtk-rs/gtk3-rs/pull/804)
 * [Update wayland-client version to 0.30](https://github.com/gtk-rs/gtk3-rs/pull/801)

[gtk4-rs](https://github.com/gtk-rs/gtk4-rs):

 * [examples/glium: lookup libepoxy-0.dll first](https://github.com/gtk-rs/gtk4-rs/pull/1153)
 * [book: Migrate to `glib-build-tools`](https://github.com/gtk-rs/gtk4-rs/pull/1024)
 * [Base Dockerfile on gtk-rs-core image](https://github.com/gtk-rs/gtk4-rs/pull/1029)
 * [image: Rebuild once every week](https://github.com/gtk-rs/gtk4-rs/pull/1162)

>  * [gtk: Generate new v4.10 APIs](https://github.com/gtk-rs/gtk4-rs/pull/1155)

Reviewer: # GTK4 4.10 API additions

The GTK4 bindings were updated to include all the new API introduced by the upcoming GTK 4.10 release. These new APIs are available via the `v4_10` feature flag and are still potentially subject to change if the underlying C API happens to change before the 4.10.0 release.

The minimum supported GTK version stays at 4.0.0.

 * [Fixed a typo](https://github.com/gtk-rs/gtk4-rs/pull/1182)
 * [Release 0.5.2](https://github.com/gtk-rs/gtk4-rs/pull/1184)
 * [Fix clippy lints](https://github.com/gtk-rs/gtk4-rs/pull/1176)
 * [Skip init assertion for `gdk::set_allowed_backends`](https://github.com/gtk-rs/gtk4-rs/pull/1183)
 * [book: Fix typo](https://github.com/gtk-rs/gtk4-rs/pull/1190)
 * [Added Libxml, Librsvg, Gettext and Libadwaita demo](https://github.com/gtk-rs/gtk4-rs/pull/1178)
 * [use `Into<Value>` and `Into<Variant>` where possible](https://github.com/gtk-rs/gtk4-rs/pull/1194)
 * [book: Small windows instructions fix](https://github.com/gtk-rs/gtk4-rs/pull/1196)
 * [examples: Use NoneCast trait where possible](https://github.com/gtk-rs/gtk4-rs/pull/1206)
 * [macros: Allow using re-exports of gtk](https://github.com/gtk-rs/gtk4-rs/pull/1179)
 * [gsk: Export builders module](https://github.com/gtk-rs/gtk4-rs/pull/1186)
 * [gtk4: use `impl_offset()` for calculating template child offset](https://github.com/gtk-rs/gtk4-rs/pull/1192)
 * [gtk: Fix missing version guards](https://github.com/gtk-rs/gtk4-rs/pull/1195)
 * [Add a changelog file](https://github.com/gtk-rs/gtk4-rs/pull/1211)

>  * [gtk: Cleanup template related functions](https://github.com/gtk-rs/gtk4-rs/pull/1212)

Reviewer: @bilelmoussaoui you mentioned this one in the gtk4-rs changelog. What would you write about it, why does it matter for users? :)

 * [book: Link to "Rust Atomics and Lock"](https://github.com/gtk-rs/gtk4-rs/pull/1252)
 * [Add blueprint-compiler dependency into Dockerfile](https://github.com/gtk-rs/gtk4-rs/pull/1239)
 * [no need to move in the hello world exercise](https://github.com/gtk-rs/gtk4-rs/pull/1254)
 * [Book: make memory management conclusion clearer.](https://github.com/gtk-rs/gtk4-rs/pull/1255)
 * [Use wrapper macros where possible](https://github.com/gtk-rs/gtk4-rs/pull/1256)
 * [Various improvments](https://github.com/gtk-rs/gtk4-rs/pull/1258)
 * [Allocations improvements](https://github.com/gtk-rs/gtk4-rs/pull/1259)
 * [gtk: Make use of `Value::from_type_unchecked`](https://github.com/gtk-rs/gtk4-rs/pull/1264)
 * [book: warn about precedence over pkg-config-lite](https://github.com/gtk-rs/gtk4-rs/pull/1263)
 * [Recommend (with instructions) building with `gvsbuild`](https://github.com/gtk-rs/gtk4-rs/pull/1203)
 * [book: Simplify instructions on Windows](https://github.com/gtk-rs/gtk4-rs/pull/1265)
 * [Make use of IntoGStr/IntoStrV](https://github.com/gtk-rs/gtk4-rs/pull/1262)
 * [Enable introspection for gtk](https://github.com/gtk-rs/gtk4-rs/pull/1269)

>  * [gtk4-macros: Add blueprint support](https://github.com/gtk-rs/gtk4-rs/pull/1238)

Reviewer: # Add blueprint UI definition support

[Blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/) is a new markup language for defining GTK4 UIs. The `#[template]` macro can now accept Blueprint UI definitions in addition to the old XML-based UI definitions. Doing so requires `blueprint` to be available during compilation.

Copy the example from https://github.com/gtk-rs/gtk4-rs/releases/tag/0.6.0

 * [Fix new clippy lint in generated abi and sys test files](https://github.com/gtk-rs/gir/pull/1398)
 * [analysis/special_functions: fix missing import](https://github.com/gtk-rs/gir/pull/1403)
 * [codegen: implement `From<T>` for Value for enums/flags](https://github.com/gtk-rs/gir/pull/1402)
 * [Fix up special functions in traits correctly](https://github.com/gtk-rs/gir/pull/1404)
 * [Fix incorrect 'missing from library' warning when generating in sys mode](https://github.com/gtk-rs/gir/pull/1405)
 * [Add init assertions to the enum/flags `From<T> for glib::Value` impls](https://github.com/gtk-rs/gir/pull/1406)
 * [codegen: silence deprecation warnings in impls for deprecated types/functions](https://github.com/gtk-rs/gir/pull/1408)
 * [analysis: handle C array pointer casts to void](https://github.com/gtk-rs/gir/pull/1407)
 * [fix the layout tests](https://github.com/gtk-rs/gir/pull/1409)
 * [codegen: Handle `_finish` functions not taking a GError inout param](https://github.com/gtk-rs/gir/pull/1397)
 * [Backport: codegen: Handle `_finish` functions not taking a GError inout param](https://github.com/gtk-rs/gir/pull/1411)

>  * [analysis: Use the new move trait for in transfer-full params](https://github.com/gtk-rs/gir/pull/1384)

Reviewer: # Better mapping of C ownership transfer

Previously various functions that took ownership of their parameters still used references in the bindings. This required an unnecessary copy to be made in the worst case, and in the best case made it less clear that ownership of the object is given away.

In this release, functions are taking various arguments by value in more places.

 * [Correct outdated references to `subclass::simple`](https://github.com/gtk-rs/gtk-rs-core/pull/808)
 * [glib: Add `ObjectSubclass::obj()` as a shorter alias for `instance()`](https://github.com/gtk-rs/gtk-rs-core/pull/806)
 * [image: Rebuild once every week](https://github.com/gtk-rs/gtk-rs-core/pull/817)
 * [Fix new clippy lints](https://github.com/gtk-rs/gtk-rs-core/pull/816)
 * [0.16 backports](https://github.com/gtk-rs/gtk-rs-core/pull/815)
 * [fix CI for 1.65/1.66](https://github.com/gtk-rs/gtk-rs-core/pull/819)
 * [Fix new clippy lints](https://github.com/gtk-rs/gtk-rs-core/pull/823)
 * [Move from `imp.instance()` to `imp.obj()`](https://github.com/gtk-rs/gtk-rs-core/pull/814)
 * [0.16 backports](https://github.com/gtk-rs/gtk-rs-core/pull/824)
 * [Move `g\_cancellable\_set\_error\_if\_cancelled()` to manual](https://github.com/gtk-rs/gtk-rs-core/pull/820)
 * [pango: Auto generate Language](https://github.com/gtk-rs/gtk-rs-core/pull/827)
 * [pango: Make `pango::Language::from\_string()` infallible](https://github.com/gtk-rs/gtk-rs-core/pull/828)
 * [Update gir files](https://github.com/gtk-rs/gtk-rs-core/pull/829)

>  * [Implement From&lt;T&gt; for Value, Variant](https://github.com/gtk-rs/gtk-rs-core/pull/826)

Reviewer: # Direct conversions to `glib::Value` and `glib::Variant` via `From`

Previously conversions required using the custom `ToValue` and `ToVariant` traits, which is not very intuitive. Now it's possible to directly use the `From` trait, i.e. `123i32.to_value()` could become `123i32.into()`.

In addition to usability this also avoid copies of the values in a few cases.

 * [glib: fix undefined behavior in types::register\_type](https://github.com/gtk-rs/gtk-rs-core/pull/836)

>  * [gdk-pixbuf: Add subclassing support](https://github.com/gtk-rs/gtk-rs-core/pull/835)

Reviewer: # `gdk_pixbuf` subclassing support

Support for implement subclasses of the various `gdk_pixbuf` types was added. This allows writing custom pixbuf loaders in Rust and hopefully opens the way to replacing the old C-based PNG/JPEG/GIF pixbuf loaders with Rust versions to improve image loading in all GTK/GNOME applications.

 * [glib: Document the value guarantees for `ObjectImpl::set\_property()` …](https://github.com/gtk-rs/gtk-rs-core/pull/832)
 * [glib: Add a doc string for `as\_ptr` generated impls](https://github.com/gtk-rs/gtk-rs-core/pull/830)
 * [0.16 backports](https://github.com/gtk-rs/gtk-rs-core/pull/837)
 * [update windows CI to glib 2.74](https://github.com/gtk-rs/gtk-rs-core/pull/838)
 * [Fix ABI tests](https://github.com/gtk-rs/gtk-rs-core/pull/840)
 * [cairo: fix some misc warnings](https://github.com/gtk-rs/gtk-rs-core/pull/841)
 * [gio: socket/stream extras](https://github.com/gtk-rs/gtk-rs-core/pull/822)
 * [cairo: Update to freetype 0.32](https://github.com/gtk-rs/gtk-rs-core/pull/844)

>  * [Add CastNone trait](https://github.com/gtk-rs/gtk-rs-core/pull/843)

Reviewer: # New `glib::CastNone` convenience trait

In many cases casting failures of `Option`al values should be treated as `None`. For this purpose a new convenience trait was introduced, which provides an `and_downcast()` method.

```rust
let widget: Option<Widget> = list_item.child();

// Without using `CastNone`
let label = widget.unwrap().downcast::<gtk::Label>().unwrap();

// Using `CastNone` we can avoid the first `unwrap()` call
let label = widget.and_downcast::<gtk::Label>().unwrap();
```

 * [glib: Implement `GStringBuilder` as `BoxedInline` to avoid a useless …](https://github.com/gtk-rs/gtk-rs-core/pull/856)
 * [glib: Minor `GStringBuilder` improvements](https://github.com/gtk-rs/gtk-rs-core/pull/857)
 * [Group imports and use prelude](https://github.com/gtk-rs/gtk-rs-core/pull/854)
 * [0.16 backports](https://github.com/gtk-rs/gtk-rs-core/pull/858)
 * [glib: Bind more `g\_utf8` APIs](https://github.com/gtk-rs/gtk-rs-core/pull/813)
 * [gio: fix clippy lints for rust 1.64](https://github.com/gtk-rs/gtk-rs-core/pull/861)
 * [gio: add spawn\_blocking()](https://github.com/gtk-rs/gtk-rs-core/pull/818)
 * [Fix various new beta clippy warnings](https://github.com/gtk-rs/gtk-rs-core/pull/862)
 * [examples: spawn async gio task on the current thread context](https://github.com/gtk-rs/gtk-rs-core/pull/863)
 * [0.16 backports](https://github.com/gtk-rs/gtk-rs-core/pull/864)
 * [Various `Stash` / `to\_glib\_none()` related optimizations and bugfixes](https://github.com/gtk-rs/gtk-rs-core/pull/865)

>  * [GString refactor](https://github.com/gtk-rs/gtk-rs-core/pull/600)

Reviewer: # `glib::GStr`, `glib::List`, `glib::PtrSlice`, `glib::Slice`, `glib::StrV` and related native GLib types

New API was added to all the beforementioned type to make usage of them more natural in Rust. Using them instead of the native Rust types, e.g. `Vec`, allows for fewer copies when passing the FFI boundary.

Various bindings were also switched to these types for the same reason, and it is planned for the next release(s) to move even more API to it.

In addition two new traits were introduced for dealing with GLib strings and string arrays more efficiently: `IntoGStr` and `IntoStrV`. Both allow passing a plain `&str` or an already `NUL`-terminated `&GStr`, or a `Vec<String>` or an already `NUL`-terminated `glib::StrV` to functions and internally the function would either do a copy or directly pass the value to the C function. Up to a certain size the copy would by done on the stack, afterwards a temporary heap copy is made. Compared to the previous copies that always happened this allows for a lot fewer temporary heap allocations.

And lastly, `glib::GString` is now storing strings up to a certain size inline without requiring a heap allocation.

 * [ActionEntry: take proper types instead of strings](https://github.com/gtk-rs/gtk-rs-core/pull/870)

>  * [Inline various trivial functions](https://github.com/gtk-rs/gtk-rs-core/pull/866)

Reviewer: # Various code generation improvements

All over the bindings, various small code generation improvements were applied which reduce the size of the generated code and/or the performance.

For example:
 - Various trivial functions were explicitly marked as `#[inline]` to give the compiler yet another hint to actually line them, which then causes the bindings to be completely optimized away in most cases.
 - In many places, assertions were converted to `debug_assert!`s if that made sense to reduce the number of place where stack unwinding can happen in release builds
 - The representation of various types was improved, e.g. by removing unnecessary `Option` wrapping around values or replacing unneeded references with `PhantomData` with the same lifetime.

---
layout: post
author: gtk-rs devs
title: New release
categories: [front, crates]
date: 2023-02-10 22:00:00 +0000
---

It’s now time for a new gtk-rs release! As you will notice, `gtk3-rs` projects are getting less and less attention and we intend to deprecate them in one of the future releases. Therefore, we recommend to anyone who didn't upgrade to GTK4 to do it.

You can also take a look at the CHANGELOG of the `gstreamer-rs` 0.20 release [here](https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/blob/main/gstreamer/CHANGELOG.md#0200-2023-02-10).

One this note, time to go through the major changes of this release. Enjoy!

## Update of minimum supported Rust version (MSRV)

With this release, Rust 1.64 is the minimum version required to compile and use the bindings.

## New `#[properties]` macro for making usage of properties in `glib::Object` subclasses easier

The `glib::Properties` macro is now available! It can be used to define object properties from a struct definition.
The macro will:
  - define all the necessary ParamSpecs, with the required flags

>   - generate, for each property, a default getter/setter on the wrapper type

Reviewer 2: On the wrapper type or on the implementation type? Or on both?

Author: What is the implementation type?

Reviewer 2: Sorry, I meant on imp::Foo or on super::Foo or on both?

Author: No idea. cc @sdroege

Reviewer: On the wrapper type. `super::Foo`. What part is confusing here? :)

Reviewer 2: The confusing part is usually you'll have getters and setters on `imp::Foo` doing the actual work (since it has access to the struct members which you need to get/set), and you'll have dummy getters and setters on `super::Foo` dispatching to `imp::Foo`. Is this somehow different with the properties macro?

Reviewer 2: Okay, understood, thanks

  - generate, for each property, a `notify_$property` method on the wrapper type

>   - generate the methods `derived_properties` `derived_property` and `derived_set_property`, which can be directly called inside the respective `ObjectImpl` methods.

Reviewer 2: ```suggestion
  - generate the methods `derived_properties` `derived_property` and `derived_set_property`, which can be directly called inside your overrides for `ObjectImpl` methods `properties`, `property` and `set_property`.
```
Wasn't clear at all to me until I read the documentation example.

### Quick example

```
#[derive(Properties)]
#[properties(wrapper_type = super::Author)]
pub struct Author {

>     #[property(get, set, explicit_notify)]

Reviewer 2: What is explicitly notifying in this case? When writing properties manually, I set explicit_notify, and then manually call notify() in set_foo() only if the new value is different from the old one. What's happening here?

Author: It's the name of the property from what I can infer.

Reviewer 2: No, as in, does the generated code call notify() in this case? Does it check for value equality, or call unconditionally?

Author: Absolutely no idea.

cc @sdroege

Reviewer: See the C docs:

> G_PARAM_EXPLICIT_NOTIFY | calls to g_object_set_property() for this property will not automatically result in a "notify" signal being emitted: the implementation must call g_object_notify() themselves in case the property actually changes. Since: 2.42.

or our docs: http://gtk-rs.org/gtk-rs-core/stable/latest/docs/glib/struct.ParamFlags.html#associatedconstant.EXPLICIT_NOTIFY

Reviewer 2: I know how explicit notify works. I'm asking how the properties macro works. Is it
```rust
fn set_foo(&self, value: i32) {
    if self.value.get() == value {
        return;
    }
    self.value.set(value);
    self.obj.notify("value");
}
```
or
```rust
fn set_foo(&self, value: i32) {
    self.value.set(value);
    self.obj.notify("value");
}
```
?

Reviewer: Neither of those. You have to explicitly notify, it does not notify automatically. It simply sets that flag on the pspec and nothing else, as it should.

There is no such magic, but that could be added at a later time under a different name to avoid exactly such questions. The `explicity_notify` flag does exactly what it does in C and in any other context.

Reviewer 2: Does it mean that this example is wrong then? Since the default `set` in `#[property(get, set, explicit_notify)]` won't actually notify the property, and you cannot easily know that it changed without providing your own `set`?

    name: RefCell<String>,
}
```

> This is the result of months of work, and will greatly improve the gtk-rs ecosystem, by reducing the boilerplate and making the platform more welcoming. 

Reviewer 2: ```suggestion
This is the result of months of work, and will greatly improve the gtk-rs ecosystem by reducing some of the most prevalent boilerplate and making the platform more welcoming. 
```
I'd say props boilerplate is annoying enough to deserve this :)

> Be sure to checkout the [documentation](https://gtk-rs.org/gtk-rs-core/stable/latest/docs/glib/derive.Properties.html) for more info.

Reviewer 2: ```suggestion
Be sure to check out the [documentation](https://gtk-rs.org/gtk-rs-core/stable/latest/docs/glib/derive.Properties.html) for more info.
```

## Direct conversions to `glib::Value` and `glib::Variant` via `From`

Previously conversions required using the custom `ToValue` and `ToVariant` traits, which is not very intuitive. Now it's possible to directly use the `From` trait, i.e. `123i32.to_value()` could become `123i32.into()`.

> In addition to usability this also avoid copies of the values in a few cases.

Reviewer 2: ```suggestion
In addition to improving usability this also avoids copies of the values in a few cases.
```

## `gdk_pixbuf` subclassing support

> Support for implement subclasses of the various `gdk_pixbuf` types was added. This allows writing custom pixbuf loaders in Rust and hopefully opens the way to replacing the old C-based PNG/JPEG/GIF pixbuf loaders with Rust versions to improve image loading in all GTK/GNOME applications.

Reviewer 2: ```suggestion
Support for subclassing various `gdk_pixbuf` types was added. This allows writing custom pixbuf loaders in Rust and hopefully opens the way for replacing the old C-based PNG/JPEG/GIF pixbuf loaders with Rust versions to improve image loading in all GTK/GNOME applications.
```

## New `glib::CastNone` convenience trait

In many cases casting failures of `Option`al values should be treated as `None`. For this purpose a new convenience trait was introduced, which provides an `and_downcast()` method.

```rust
let widget: Option<Widget> = list_item.child();

// Without using `CastNone`
let label = widget.unwrap().downcast::<gtk::Label>().unwrap();

// Using `CastNone` we can avoid the first `unwrap()` call
let label = widget.and_downcast::<gtk::Label>().unwrap();
```

## New `glib::function_name!` macro

> While a huge hack, until Rust supports this directly this macro allows to get the name of the current function as a `&'static str`. Or to be more exact, the whole path to the current function.

Reviewer 2: ```suggestion
While a huge hack, until Rust supports this directly, this macro will allow to get the name of the current function as a `&'static str`. Or to be more exact, the whole path to the current function.
```

This is now used in the GLib structured logging API to provide the function name (in addition to the file name, line number and other metadata) and `glib::BoolError` to provide better location information.

## `glib::GStr`, `glib::List`, `glib::PtrSlice`, `glib::Slice`, `glib::StrV` and related native GLib types

> New API was added to all the beforementioned type to make usage of them more natural in Rust. Using them instead of the native Rust types, e.g. `Vec`, allows for fewer copies when passing the FFI boundary.

Reviewer 2: ```suggestion
New API was added to all the aforementioned types to make dealing with them more natural in Rust. Using them instead of the native Rust types, e.g. `Vec`, allows for fewer copies when passing the FFI boundary.
```

Various bindings were also switched to these types for the same reason, and it is planned for the next release(s) to move even more API to it.

> In addition two new traits were introduced for dealing with GLib strings and string arrays more efficiently: `IntoGStr` and `IntoStrV`. Both allow passing a plain `&str` or an already `NUL`-terminated `&GStr`, or a `Vec<String>` or an already `NUL`-terminated `glib::StrV` to functions and internally the function would either do a copy or directly pass the value to the C function. Up to a certain size the copy would by done on the stack, afterwards a temporary heap copy is made. Compared to the previous copies that always happened this allows for a lot fewer temporary heap allocations.

Reviewer 2: ```suggestion
In addition two new traits were introduced for dealing with GLib strings and string arrays more efficiently: `IntoGStr` and `IntoStrV`. Both allow passing a plain `&str` or an already `NUL`-terminated `&GStr`, or a `Vec<String>` or an already `NUL`-terminated `glib::StrV` to functions and internally the function would either do a copy or directly pass the value to the C function. Up to a certain size the copy would by done on the stack, afterwards a temporary heap copy is made. Compared to the before, when temporary heap copies were made unconditionally, this allows for a lot fewer temporary heap allocations.
```

And lastly, `glib::GString` is now storing strings up to a certain size inline without requiring a heap allocation. New `GString`s can be created via the `gformat!` macro similar to the `format!` macro from `std`.

## Various code generation improvements

> All over the bindings, various small code generation improvements were applied which reduce the size of the generated code and/or the performance.

Reviewer 2: ```suggestion
All over the bindings, various small code generation improvements were applied which reduced the size of the generated code and/or improved the performance.
```
I hope it didn't *reduce* the performance :)

For example:

>  - Various trivial functions were explicitly marked as `#[inline]` to give the compiler yet another hint to actually line them, which then causes the bindings to be completely optimized away in most cases.

Reviewer 2: ```suggestion
 - Various trivial functions were explicitly marked as `#[inline]` to give the compiler yet another hint to actually inline them, and improve inlining across crate boundaries, which then causes the bindings to be completely optimized away in most cases.
```

>  - In many places, assertions were converted to `debug_assert!`s if that made sense to reduce the number of place where stack unwinding can happen in release builds

Reviewer 2: ```suggestion
 - In many places, where it made sense, assertions were converted to `debug_assert!`s to reduce the number of places where stack unwinding could happen in release builds.
```

 - The representation of various types was improved, e.g. by removing unnecessary `Option` wrapping around values or replacing unneeded references with `PhantomData` with the same lifetime.

>  - Using uninitialized memory for C functions to fill instead of zero-initializing it

Reviewer 2: ```suggestion
 - Memory initialized by C functions was converted from using zero-initialization to `MaybeUninit`.
```
I hope you're using MaybeUninit and not the unsafe `mem::uninitialized` :)

>  - Reducing the number of redundant checks and adding `unsafe` variants of functions that skip the checks

Reviewer 2: ```suggestion
 - The number of redundant checks was reduced and `unsafe` variants of functions that skip the checks were added.
```

>  - All object builder types were changed to use `glib::Object::builder()`, which reduces stack usage dramatically in addition to requiring fewer value copies

Reviewer 2: ```suggestion
 - All object builder types were changed to use `glib::Object::builder()`, which reduces stack usage dramatically in addition to requiring fewer value copies.
```

## New `glib::Object` constructors

> The old `Object::new()` constructor that took a slice of property name / value pairs does not exist anymore. Instead `Object::new()` takes no parameters anymore and creates a new object with the default property values.

Reviewer 2: ```suggestion
The old `Object::new()` constructor that took a slice of property name / value pairs was removed. Instead, `Object::new()` takes no parameters and creates a new object with default property values.
```

> For creating a new object with non-default property values, `Object::builder()` can be used which provides a much nicer API and also allows for better code generation and fewer value copies.

Reviewer 2: ```suggestion
For creating a new object with non-default property values, use `Object::builder()`. It provides a much nicer API and also allows for better code generation and fewer value copies.
```

```rust
// before
let obj = glib::Object::new::<MyObject>(&[("a", &123i32), ("b", &"some string")]);
// after
let obj = glib::Object::builder::<MyObject>()
    .property("a", 123i32)
    .property("b", "some string")
    .build();
```

## `gio::Application::run()` returns an `ExitCode`

> Instead of returning a plain `i32`, `gio::Application::run()` now returns an `ExitCode` type. This can be directly returned from the `main()` function as it implements the `Termination` trait and by that makes it much easier to correctly propagate application exit codes.

Reviewer 2: ```suggestion
Instead of returning a plain `i32`, `gio::Application::run()` now returns an `ExitCode` type. This can be directly returned from the `main()` function as it implements the `Termination` trait, making it much easier to correctly propagate application exit codes.
```

## GTK4 4.10 API additions

> The GTK4 bindings were updated to include all the new API introduced by the upcoming GTK 4.10 release. These new APIs are available via the `v4_10` feature flag and are still potentially subject to change if the underlying C API happens to change before the 4.10.0 release.

Reviewer 2: ```suggestion
The GTK4 bindings were updated to include all the new APIs introduced by the upcoming GTK 4.10 release. These new APIs are available via the `v4_10` feature flag. Note that they are **not stable** and are subject to change if the underlying C API happens to change before the 4.10.0 release.
```
I feel like this deserves more emphasis

The minimum supported GTK version stays at 4.0.0.

> ## Add blueprint UI definition support

Reviewer 2: ```suggestion
## Add Blueprint UI definition support
```

> [Blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/) is a new markup language for defining GTK4 UIs. The `#[template]` macro can now accept Blueprint UI definitions in addition to the old XML-based UI definitions. Doing so requires `blueprint` to be available during compilation.

Reviewer 2: ```suggestion
[Blueprint](https://jwestman.pages.gitlab.gnome.org/blueprint-compiler/) is a new markup language for defining GTK4 UIs. The `#[template]` macro can now accept Blueprint UI definitions in addition to the old XML-based UI definitions. Doing so requires `blueprint-compiler` to be available during compilation.
```

> Copy the example from https://github.com/gtk-rs/gtk4-rs/releases/tag/0.6.0

Reviewer 2: ```rust
#[derive(Debug, Default, gtk::CompositeTemplate)]
#[template(string = "
template MyWidget : Widget {
    Label label {
        label: 'foobar';
    }
    Label my_label2 {
        label: 'foobaz';
    }
}
")]
pub struct MyWidget {
    #[template_child]
    pub label: TemplateChild<gtk::Label>,
    #[template_child(id = "my_label2")]
    pub label2: gtk::TemplateChild<gtk::Label>,
}
```
I could not figure out how to suggest with inline code

Author: Should I just copy this code directly into the blog post then?

Reviewer 2: Yeah, I don't see why not.

## Better mapping of C ownership transfer

Previously various functions that took ownership of their parameters still used references in the bindings. This required an unnecessary copy to be made in the worst case, and in the best case made it less clear that ownership of the object is given away.

In this release, functions are taking various arguments by value in more places.

## Spawning of blocking functions on the `glib::MainContext`

> `gio::spawn_blocking()` was added, which allows to spawn blocking functions on an internal threadpool and retrieve their results as a `Future` from the `glib::MainContext`. This is useful for doing blocking / CPU intense work in the background and when the results are available to handle them on the application's main thread.

Reviewer 2: ```suggestion
`gio::spawn_blocking()` was added, which allows to spawn blocking functions on an internal threadpool and retrieve their results as a `Future` from the `glib::MainContext`. This is useful for doing blocking / CPU intensive work in the background and when the results are available to handle them on the application's main thread.
```

As part of this work, `JoinHandle`-style types were also added to `glib::MainContext::spawn()` and related functions for spawning `Future`s on the main context, and to `glib::ThreadPool` for spawning functions on a custom thread pool and collecting their results.

# Changes

> For the interested ones, here is the list of the merged pull requests:

Reviewer 2: ```suggestion
For those who are interested, here is the list of the merged pull requests:
```
